From 9aef685091d6157e378fa1f27dd9d0b3781d419f Mon Sep 17 00:00:00 2001
From: Tao Liu <taoliu828@163.com>
Date: Tue, 14 Mar 2023 20:10:17 +0800
Subject: [PATCH 1342/1646] [Backport] skbuff: Fix nfct leak on napi stolen

stable inclusion
from stable-v5.10.175
commit c53d50d8081a49ba21f866a51277a012b9efad8e
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6XGEC

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=c53d50d8081a49ba21f866a51277a012b9efad8e

--------------------------------

Upstream commit [0] had fixed this issue, and backported to kernel 5.10.54.
However, nf_reset_ct() added in skb_release_head_state() instead of
napi_skb_free_stolen_head(), which lead to leakage still exist in 5.10.

[0] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=8550ff8d8c75416e984d9c4b082845e57e560984

Fixes: 570341f10ecc ("skbuff: Release nfct refcount on napi stolen or re-used skbs"))
Signed-off-by: Tao Liu <taoliu828@163.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Wang Hai <wanghai38@huawei.com>
---
 net/core/dev.c    | 1 +
 net/core/skbuff.c | 1 -
 2 files changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index c1aa94b013dc..01569b48f02f 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -6116,6 +6116,7 @@ EXPORT_SYMBOL(gro_find_complete_by_type);
 
 static void napi_skb_free_stolen_head(struct sk_buff *skb)
 {
+	nf_reset_ct(skb);
 	skb_dst_drop(skb);
 	skb_ext_put(skb);
 	kmem_cache_free(skbuff_head_cache, skb);
diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index 9184a82e2739..f9baa52c8ce3 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -675,7 +675,6 @@ static void kfree_skbmem(struct sk_buff *skb)
 
 void skb_release_head_state(struct sk_buff *skb)
 {
-	nf_reset_ct(skb);
 	skb_dst_drop(skb);
 	if (skb->destructor) {
 		WARN_ON(in_irq());
-- 
2.17.1

